test/infrastructure/docker/internal/docker/kind_manager.go

```diff
@@ -55,21 +49,21 @@ type nodeCreateOpts struct {
 }
 
 // CreateControlPlaneNode will create a new control plane container.
-func (m *Manager) CreateControlPlaneNode(ctx context.Context, name, image, clusterName, listenAddress string, port int32, mounts []v1alpha4.Mount, portMappings []v1alpha4.PortMapping, labels map[string]string, ipFamily clusterv1.ClusterIPFamily) (*types.Node, error) {
+func (m *Manager) CreateControlPlaneNode(ctx context.Context, name, image, clusterName, listenAddress string, hostPort, cpPort int32, mounts []v1alpha4.Mount, portMappings []v1alpha4.PortMapping, labels map[string]string, ipFamily clusterv1.ClusterIPFamily) (*types.Node, error) {
 	// gets a random host port for the API server
-	if port == 0 {
+	if hostPort == 0 {
 		p, err := getPort()
 		if err != nil {
 			return nil, errors.Wrap(err, "failed to get port for API server")
 		}
-		port = p
+		hostPort = p
 	}
 
 	// add api server port mapping
 	portMappingsWithAPIServer := append(portMappings, v1alpha4.PortMapping{
 		ListenAddress: listenAddress,
-		HostPort:      port,
-		ContainerPort: KubeadmContainerPort,
+		HostPort:      hostPort,
+		ContainerPort: cpPort,
```

Comment:
Sorry probably a bit late for this question. What ports are we actually changing with this PR?

It looks to me like:
* we are changing the internal port of the haproxy (while the hostPort is still random)
* we are also changing this part mapping here which means the still random hostPort is now mapping to a changed internal port? (but what is listening on this port?)

I"m probably misinterpreting a few things. What's your point of view looking at haproxy (including its config) + control plane nodes + their corresponding port mappings? Which ports through the entire chain are now controlled with the DockerCluster controlPlane endpoint port setting?

Reply:
RKE2 has to open a specific port for cluster initialization: 9345. This port is used by the first control plane node, called server, to listen to new node registrations. 
The way CAPD is implemented right now prevents our bootstrap provider to register new nodes after the first one. CAPD would need to be configurable for the ports that need to be open.

Reply:
I was thinking about it more, I would prefer to only configure the frontend haproxy port. Seems like changing the port mapping on the control plane node might not be necessary.

Reply:
@belgaied2 Actually we also want to be able to append additional haproxy configuration here https://github.com/kubernetes-sigs/cluster-api/blob/main/test/infrastructure/docker/internal/third_party/forked/loadbalancer/config.go#L58
According to the [docs](https://docs.rke2.io/install/quickstart#3-configure-the-rke2-agent-service), the RKE2 server agent runs simultaneously with the k8s API server. This means we need to expose 2 ports - 9345 and 6443.

After more investigation, this PR makes less sense for RKE2 provider but the `spec.ControlPlaneEnpoint` field is still not usable in CAPD.

Reply:
I'm absolutely open to useful improvements. I just realized that I don't really understand which port or port mapping we now can configure :) 

Some sort of illustration would be great to understand what the plan is, e.g.

```
Components: kubectl (e.g.)  =>    ha proxy (lb)                   =>   backends (control plane nodes)
Ports & port mappings:                  hostport : port                       hostport : apiserver port
Example (main version):                 39433 (random) : 6443       40343(random) : 6443
Corresponding API fields:
```
(not sure if it's a good/correct way to illustrate it)

Bonus info, the haproxy config today uses a config like this:
```
frontend control-plane
  bind *:6443

  default_backend kube-apiservers
backend kube-apiservers
  ...
  server capi-quickstart-control-plane-jnm6j 172.18.0.5:6443 check check-ssl verify none
```
(which means it only uses "internal" ports and `172.18.0.5` is the IP I can see in `docker inspect` and **not** the internal Node IP)

I think usually controlPlaneEndpoint host/port is what stands for the endpoint reachable by consumers (kubectl, ...).
So it's probably good to keep this consistent. If there are reasons to change other ports we can discuss them and probably find a way to make them configurable as well.



(cc @fabriziopandini)

Reply:
tl;dr

> I was thinking about it more, I would prefer to only configure the frontend haproxy port. Seems like changing the port mapping on the control plane node might not be necessary.
I agree (at least it seems like that's what the controlPlaneEndpoint should control)

Reply:
I have commented on https://github.com/kubernetes-sigs/cluster-api/issues/7684 about using the CAPD load balancer for other services different than the controlPlane load balancer. 

If you want we can keep this PR for making the only the load balancer port configurable or close it, I live this to you

Reply:
@fabriziopandini @sbueringer I pushed some changes here, now only lb port is configurable